src/rust/inetstack/protocols/layer4/tcp/established/receiver.rs

```diff
@@ -505,40 +534,42 @@ impl Receiver {
         Ok(())
     }
 
-    fn process_data(
+    fn process_data_if_any(
         cb: &mut ControlBlock,
         layer3_endpoint: &mut SharedLayer3Endpoint,
         data: DemiBuffer,
         seg_start: SeqNumber,
         seg_end: SeqNumber,
         seg_len: u32,
     ) -> Result<(), Fail> {
-        // We can only process in-order data.  Check for out-of-order segment.
-        if seg_start != cb.receiver.receive_next_seq_no {
-            debug!(
-                "Received out-of-order segment; out_of_order_frames.len() = {:?}",
-                cb.receiver.out_of_order_frames.len()
-            );
-            debug_assert_ne!(seg_len, 0);
-            // This segment is out-of-order.  If it carries data, we should store it for later processing
-            // after the "hole" in the sequence number space has been filled.
-            match cb.state {
-                State::Established | State::FinWait1 | State::FinWait2 => {
-                    debug_assert_eq!(seg_len, data.len() as u32);
-                    cb.receiver.store_out_of_order_segment(seg_start, seg_end, data);
-                    // Sending an ACK here is only a "MAY" according to the RFCs, but helpful for fast retransmit.
-                    trace!("process_data(): send ack on out-of-order segment");
-                    Sender::send_ack(cb, layer3_endpoint);
-                },
-                state => warn!("Ignoring data received after FIN (in state {:?}).", state),
+        if data.len() > 0 {
```

Comment:
We should return early by flipping the condition. This will reduce the level of nesting in this function as well.